### DISCUSSION TITLE
Deployment Issues with Node.js Application on Vercel

### DISCUSSION THREAD

**Discussion by KabaTubare** [vercel/community#4875] [Help] [answered]
### Summary

I am experiencing persistent deployment issues with my Node.js application on Vercel and need assistance in resolving them. Below, I've outlined the key points of the problem, what I've attempted so far, and some specific queries that I'm hoping you can help me with.

Issue Summary:
I'm trying to deploy a Node.js application that interacts with the OpenAI API to retrieve stock prices. The deployment process completes without errors, but when making POST requests to the /chat-completion endpoint, I'm consistently receiving 401 Unauthorized errors. Additionally, GET requests to my root URL result in 404 Not Found errors.

Troubleshooting Steps Taken:

Verified that the API keys are correctly stored and referenced in Vercel's environment variable settings.
Ensured that the .env file is not being used in the production environment.
Checked that the index.js file is set up to handle POST requests to the /chat-completion endpoint.
Confirmed that the Node.js version I'm using locally matches the version used on Vercel (Node.js 18.x).
Deployment Details:

Deployment URL: https://my-stock-tips-project-9joq20e3j-kabatubare.vercel.app/
Vercel Project Name: my-stock-tips-project
The most recent deployment logs indicate a series of 401 and 404 errors.
Environment variables for API keys are set via the Vercel dashboard.
Code and Configuration Files:

My index.js file is configured to serve an Express.js application and listens on the appropriate port.
vercel.json is set up with the following routing configuration:
json
Copy code
{
  "version": 2,
  "builds": [{ "src": "index.js", "use": "@vercel/node" }],
  "routes": [{ "src": "/.*", "dest": "/index.js" }],
  "env": {
    "OPENAI_API_KEY": "@openai_api_key",
    "YH_FINANCE_API_KEY": "@yh_finance_api_key"
  }
}
Questions and Requests:

Could you provide insights into why the 401 Unauthorized errors are occurring during POST requests, despite the API keys being correctly configured in the environment variables?
What could be the reason for 404 Not Found errors on the root URL when the routing seems to be configured correctly in vercel.json?
Are there any additional logs or diagnostics available on your end that could shed light on these issues?
Is there any common misconfiguration I might have overlooked that typically leads to these errors?
Can you confirm if there are any additional settings or configurations on the Vercel platform that I need to adjust to resolve these issues?
I appreciate any guidance you can provide and am willing to supply any further information required to troubleshoot this issue.

Thank you for your time and assistance.

Kindly
TWoods

### Steps to Reproduce

Description:
I'm experiencing issues with deploying a Node.js serverless function on Vercel that is supposed to interact with OpenAI and fetch stock prices. When attempting to make a POST request to my endpoint, I receive a 401 Unauthorized error, and when trying to access the route through a browser, I get a 404 Not Found error.

Steps to Reproduce:

Push the latest code to my Git repository linked with the Vercel project.
Vercel triggers an automatic deployment.
Once deployed, I use the curl command to make a POST request:
css
Copy code
curl -X POST 'https://my-stock-tips-project-9joq20e3j-kabatubare.vercel.app/chat-completion' -H 'Content-Type: application/json' -d '{"message":"What is the stock price of AAPL?"}'
This returns an HTML page with a title "Authentication Required".
Accessing the endpoint through a web browser at https://my-stock-tips-project-9joq20e3j-kabatubare.vercel.app/chat-completion results in a 404 error.
Additional Information:

The deployment URL I am using is https://my-stock-tips-project-9joq20e3j-kabatubare.vercel.app.
My vercel.json configuration is as follows:
json
Copy code
{
   "version": 2,
   "builds": [
      { "src": "index.js", "use": "@vercel/node" }
   ],
   "routes": [
      { "src": "/.*", "dest": "/index.js" }
   ],
   "env": {
      "OPENAI_API_KEY": "@openai_api_key",
      "YH_FINANCE_API_KEY": "@yh_finance_api_key"
   }
}
Environment variables are set up in the Vercel project settings.
I have verified that the Node.js version on Vercel matches my local development environment (Node.js 18.x).
Attempted Solutions:

I have double-checked API keys and environment variable configurations.
I have ensured that the .env file is not being used in production.
I have confirmed that index.js is set up to handle POST requests to /chat-completion.
I have reviewed the Node.js version for compatibility with Vercel’s runtime environment.
Error Messages and Logs:

401 Unauthorized error upon POST request via curl.
404 Not Found error when accessing the endpoint via a web browser.
Deployment logs show several 404 errors for GET requests and 401 errors for POST requests to /chat-completion.

**Comment by amyegan** [comment, marked as answer]
Since the homepage of your app is returning a 404 in the browser, I suspect something has gone wrong with the deployment. Did you follow the Express.js with Vercel guide? Or is your project set up a different way?

https://vercel.com/guides/using-express-with-vercel

**Reply by chinmaym505** [reply]
I am having a similar issue, and I did look at the guide, but I did not understand it.

**Comment by Krisuv16** [comment]
Hey guy, might be a bit helpful, might not be as well. I had to put all the files i had in an API folder. first of all. And in my vercel.json i just did the default.
{ "version": 2, "rewrites": [{ "source": "/(.*)", "destination": "/api" }] }.

///////
I did try adding more stuffs like this
{
"version": 2,
"builds": [
{ "src": "index.js", "use": "@vercel/node" }
],
"routes": [
{ "src": "/.*", "dest": "/index.js" }
],
}
//////

I did add all the stuffs from my .env inside the Vercel build setup.

This was my folder structure. 
<img width="165" alt="Screenshot 2024-03-12 at 21 40 51" src="https://github.com/vercel/community/assets/48934328/473b4f72-1609-4898-a0a1-343a1d299cda">
